File: R/liquid_chart.R
#' Create a liquid fill gauge chart.
#'
#' @description
#' A liquid fill gauge represents a single value as a circle filled with an
#' animated liquid up to the corresponding level. The fill rises smoothly on
#' load, two wave layers give the water a natural depth, and the label text
#' changes color at the water surface — appearing in \code{textColor} above
#' the water and in \code{waveTextColor} below it.
#'
#' @param value Numeric. The fill level, between 0 and 1 (0 = empty, 1 = full).
#' @param label Character. Text displayed in the center of the gauge. Defaults
#' to \code{value} formatted as a percentage (e.g. \code{"55\%"}).
#' @param fillColor The color of the liquid. Defaults to \code{"steelblue"}.
#' @param circleColor The color of the outer ring. Defaults to
#' \code{fillColor}.
#' @param circleThickness Thickness of the outer ring as a fraction of the
#' circle radius. Defaults to \code{0.05}.
#' @param textColor Color of the label text above the water surface. Defaults
#' to \code{fillColor}.
#' @param waveTextColor Color of the label text below (inside) the water
#' surface. Defaults to \code{"white"}.
#' @param textSize Font size of the label as a fraction of the circle radius.
#' Defaults to \code{0.38}.
#' @param waveAmplitude Height of the wave as a fraction of the circle radius.
#' Defaults to \code{0.09}.
#' @param waveCount Number of full wave cycles visible across the circle.
#' Defaults to \code{2}.
#' @param waveSpeed Speed of the horizontal wave animation. Higher values are
#' faster. Defaults to \code{1}.
#' @param title Optional. A label displayed below the circle.
#' @param titleFontSize Font size of the title in pixels. Defaults to \code{14}.
#' @param font The font family used for all text. Defaults to
#' \code{"Verdana, Geneva, Tahoma, sans-serif"}.
#' @param bgcol Background color of the SVG. Defaults to \code{"white"}.
#' @param width Optional. The width of the SVG output.
#' @param height Optional. The height of the SVG output.
#'
#' @return An animated SVG liquid fill gauge.
#' @export
#'
#' @examples
#' # Basic usage — 55% fill
#' liquid_chart(value = 0.55)
#'
#' # Custom color and label
#' liquid_chart(
#'   value     = 0.72,
#'   label     = "72%",
#'   fillColor = "#e67e22",
#'   title     = "Disk usage"
#' )
#'
#' # Low fill with a raw value label
#' liquid_chart(
#'   value         = 0.28,
#'   label         = "28%",
#'   fillColor     = "tomato",
#'   circleColor   = "tomato",
#'   textColor     = "tomato",
#'   waveTextColor = "white"
#' )
liquid_chart <- function(
    value,
    label         = NULL,
    fillColor     = "steelblue",
    circleColor   = NULL,
    circleThickness = 0.05,
    textColor     = NULL,
    waveTextColor = "white",
    textSize      = 0.38,
    waveAmplitude = 0.09,
    waveCount     = 2,
    waveSpeed     = 1,
    title         = NULL,
    titleFontSize = 14,
    font          = "Verdana, Geneva, Tahoma, sans-serif",
    bgcol         = "white",
    width         = NULL,
    height        = NULL
) {

  if (!is.numeric(value) || length(value) != 1) {
    stop("'value' must be a single numeric value between 0 and 1.")
  }

  value <- min(max(value, 0), 1)

  if (is.null(label)) {
    label <- paste0(round(value * 100, 1), "%")
  }

  if (is.null(circleColor)) circleColor <- fillColor
  if (is.null(textColor))   textColor   <- fillColor

  if (grepl(";", font)) {
    stop("please remove the ';' character from your font argument")
  }

  r2d3::r2d3(
    data   = value,
    script = system.file("d3/scatterplot/liquidchart.js", package = "ddplot"),
    options = list(
      label           = label,
      fillColor       = fillColor,
      circleColor     = circleColor,
      circleThickness = circleThickness,
      textColor       = textColor,
      waveTextColor   = waveTextColor,
      textSize        = textSize,
      waveAmplitude   = waveAmplitude,
      waveCount       = waveCount,
      waveSpeed       = waveSpeed,
      title           = title,
      titleFontSize   = titleFontSize,
      font            = font,
      bgcol           = bgcol
    ),
    width  = width,
    height = height
  )
}

File: inst/d3/scatterplot/liquidchart.js
const fillLevel     = Math.min(Math.max(data, 0), 1);
const label         = options.label         || (Math.round(fillLevel * 1000) / 10) + "%";
const fillColor     = options.fillColor     || "steelblue";
const circleColor   = options.circleColor   || fillColor;
const circleThickR  = options.circleThickness != null ? options.circleThickness : 0.05;
const textColor     = options.textColor     || fillColor;
const waveTextColor = options.waveTextColor || "white";
const textSizeR     = options.textSize      != null ? options.textSize : 0.38;
const waveAmplR     = options.waveAmplitude != null ? options.waveAmplitude : 0.09;
const waveCount     = options.waveCount     || 2;
const waveSpeed     = options.waveSpeed     || 1;
const title         = options.title         || null;
const titleSize     = options.titleFontSize || 14;
const fontFamily    = options.font    || "Verdana, Geneva, Tahoma, sans-serif";
const bgcol         = options.bgcol   || "white";

svg.selectAll("*").remove();
svg.style("background-color", bgcol);

// --- Geometry ---
const titleH   = title ? titleSize + 10 : 0;
const cx       = width  / 2;
const cy       = (height - titleH) / 2;
const radius   = Math.min(width, height - titleH) / 2 * 0.82;
const circleThick = radius * circleThickR;
const innerR   = radius - circleThick;
const waveH    = innerR * waveAmplR;
const waveLength = (innerR * 2) / waveCount;     // px per wave cycle
const textSize = innerR * textSizeR;

// Target fill Y (SVG y=0 is top, so filled from bottom)
const targetFillY = cy + innerR - fillLevel * 2 * innerR;

// --- Unique IDs (safe for multiple charts on same page) ---
const uid         = Math.random().toString(36).slice(2, 8);
const circleClipId = "liq-circle-" + uid;
const waveClipId   = "liq-wave-"   + uid;

const defs = svg.append("defs");

// Circle clip path
defs.append("clipPath").attr("id", circleClipId)
  .append("circle")
  .attr("cx", cx).attr("cy", cy).attr("r", innerR - 1);

// Wave clip path (updated each frame) — defines the "inside water" region
const waveClipGroup = defs.append("clipPath").attr("id", waveClipId);
const waveClipPath  = waveClipGroup.append("path");

// --- Background tint (empty part of circle) ---
svg.append("circle")
  .attr("cx", cx).attr("cy", cy).attr("r", innerR - 1)
  .attr("fill", fillColor)
  .attr("opacity", 0.1);

// --- text1: above-water label (behind waves, normal textColor) ---
const text1 = svg.append("text")
  .attr("x", cx).attr("y", cy)
  .attr("text-anchor", "middle")
  .attr("dominant-baseline", "middle")
  .attr("font-size",   textSize)
  .attr("font-family", fontFamily)
  .attr("font-weight", "bold")
  .attr("fill", textColor)
  .text(label);

// --- Fill group (wave visuals, clipped to circle) ---
const fillGroup = svg.append("g").attr("clip-path", `url(#${circleClipId})`);
const wave1     = fillGroup.append("path").attr("fill", fillColor).attr("opacity", 0.9);
const wave2     = fillGroup.append("path").attr("fill", fillColor).attr("opacity", 0.45);

// --- Circle outline (drawn on top of fill so edge is crisp) ---
svg.append("circle")
  .attr("cx", cx).attr("cy", cy).attr("r", radius)
  .attr("fill",         "none")
  .attr("stroke",       circleColor)
  .attr("stroke-width", circleThick);

// --- text2: below-water label (clipped to wave area, waveTextColor) ---
svg.append("text")
  .attr("clip-path",         `url(#${waveClipId})`)
  .attr("x", cx).attr("y", cy)
  .attr("text-anchor",       "middle")
  .attr("dominant-baseline", "middle")
  .attr("font-size",         textSize)
  .attr("font-family",       fontFamily)
  .attr("font-weight",       "bold")
  .attr("fill",              waveTextColor)
  .text(label);

// --- Wave path generator ---
// Draws from x = cx-2*innerR to cx+2*innerR (wider than circle for seamless loop)
// fillY: current y of the still-water surface
// phase: horizontal phase offset in radians
function makePath(fillY, phase) {
  const xL  = cx - innerR * 2.2;
  const xR  = cx + innerR * 2.2;
  const bot = cy + innerR + 4;
  const step = (xR - xL) / 64;   // 64 segments — smooth but fast

  let d = `M ${xL.toFixed(1)},${bot}`;
  for (let x = xL; x <= xR; x += step) {
    const y = fillY + waveH * Math.sin(((x - cx) / waveLength) * 2 * Math.PI + phase);
    d += ` L ${x.toFixed(1)},${y.toFixed(1)}`;
  }
  d += ` L ${xR.toFixed(1)},${bot} Z`;
  return d;
}

// --- Animation ---
const riseDuration = 1400; // ms for water to rise to target level
let startTime = null;

d3.timer(elapsed => {
  if (startTime === null) startTime = elapsed;
  const dt = elapsed - startTime;

  // Water-rise easing (stops after riseDuration)
  const riseT    = Math.min(1, dt / riseDuration);
  const easedT   = 1 - Math.pow(1 - riseT, 3); // ease-out cubic
  const fillY    = cy + innerR - easedT * fillLevel * 2 * innerR;

  // Continuous horizontal wave motion (phase in radians)
  const phase    = (dt / 1000) * waveSpeed * 1.8;

  // Primary wave (phase offset 0)
  const p1 = makePath(fillY, phase);
  wave1.attr("d", p1);

  // Second wave (slightly offset phase and slower — gives depth)
  const p2 = makePath(fillY + waveH * 0.3, phase * 0.7 + Math.PI * 0.6);
  wave2.attr("d", p2);

  // Keep wave clip in sync with wave1 so text2 is revealed correctly
  waveClipPath.attr("d", p1);
});

// --- Title below the circle ---
if (title) {
  svg.append("text")
    .attr("x",            cx)
    .attr("y",            cy + innerR + circleThick + titleSize + 6)
    .attr("text-anchor",  "middle")
    .attr("font-size",    titleSize)
    .attr("font-family",  fontFamily)
    .attr("fill",         fillColor)
    .text(title);
}
